### PR TITLE
Adding the ability for ivy repo's to use a non-standard pattern

### PR DESCRIPTION
**Pull request by ethankhall**
With the current implementation of the PluginRepositories it's not
possible for the user to be able to specify a repo that doens't meet the
"ivy" standard layout. Many people do not, so it's impossible to use
this feature.

This change will allow for users to specify more info about their
repositories.

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [ ] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/plugin/repository/IvyPluginRepository.java [hunk 1]
```diff
@@ -44,4 +47,103 @@
...
+    void layout(String layoutName, Action<? extends RepositoryLayout> config);
+
+    /**
+     * Specifies how the items of the repository are organized. See {@link #layout(String, org.gradle.api.Action)}
+     *
+     * @param layoutName The name of the layout to use.
+     * @param config The closure used to configure the layout.
+     * An instance of {@link RepositoryLayout} is passed as a parameter to the closure.
+     */
+    void layout(String layoutName, Closure config);
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/plugin/repository/IvyPluginRepository.java]: Closure shouldn't be necessary, added by DSL decoration.

Comment (ethankhall) [inline subprojects/core/src/main/java/org/gradle/plugin/repository/IvyPluginRepository.java]: I can remove it, do you want me to also remove it from the RepositoryHandler that it was copied from?

Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/plugin/repository/IvyPluginRepository.java]: No, we'll remove that when we remove all the old `Closure` overloads.

**Comment by oehme** [comment]
@ethankhall We're branching for the release next Monday, so in order to make it into 4.0, we need  the test coverage for this change this week.

**Comment by ethankhall** [comment]
Thanks for the notification @oehme, I've added the test and removed the method. Let me know if there's anything else you need from me.

**Pull request merged**